Text encoding speedup

lucacasonato opened the pull request:
I have sped up the TextEncoder for UTF-8 encoding by utilizing a encoding function from https://github.com/samthor/fast-text-encoding.

The speed improvement is about 4.5x. This is probably because there are way less allocations as the encoded bytes are written right into a Uint8Array. Maybe this can be optimized further, but it is a definite improvement over what we have now.

I did not do this for TextEncoder#encodeInto as I can not guarantee that passed Uint8Array() is long enough for the content. TextEncoder#encodeInto still uses the old encoder.

ry commented:
Great! Can you patch deno_website2 so that the benchmark displays and we can observe this improvement?

lucacasonato commented:
> Great! Can you patch deno_website2 so that the benchmark displays and we can observe this improvement?

Already does. The website picks up new variants for existing benchmarks automatically. As this is just a new variation for the speed benchmarks (Execution time, Thread count, Max Memory Usage, ect.) its on there already.

BTW idk why GitHub is showing that I added the perf benchmark in this PR. I added that in #3589 so the website already has one data point for that benchmarks before this speedup.

nayeemrmn commented:
> BTW idk why GitHub is showing that I added the perf benchmark in this PR. I added that in #3589 so the website already has one data point for that benchmarks before this speedup.

You've based this on your first PR commit (https://github.com/denoland/deno/pull/3589/commits/fbd4bbb126ced039f224bed5ad069898faec274a) as opposed to the one merged onto master (https://github.com/denoland/deno/commit/0a900949c83503c2af75e8f1ea9bbbc6e947887d). I don't know that it makes a difference, though.
